WalletWasabi/Wallets/FilterProcessor/BlockDownloadService.cs

```diff
@@ -0,0 +1,276 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Wallets.FilterProcessor;
+
+/// <summary>
+/// Service that downloads blocks downloads.
+/// </summary>
+public class BlockDownloadService : BackgroundService
+{
+	/// <summary>Maximum number of parallel block-downloading tasks.</summary>
+	private const int MaxParallelTasks = 5;
+
+	/// <summary>Maximum number of attempts to download a block. If it fails, we drop the block download request altogether.</summary>
+	public const int MaxFailedAttempts = 3;
+
+	public BlockDownloadService(IFileSystemBlockRepository fileSystemBlockRepository, IBlockProvider blockProvider, int maximumParallelTasks = MaxParallelTasks)
+	{
+		FileSystemBlockRepository = fileSystemBlockRepository;
+		BlockProvider = blockProvider;
+		MaximumParallelTasks = maximumParallelTasks;
+	}
+
+	/// <summary>Source that provided a block or blocks.</summary>
+	public enum Source
+	{
+		FileSystemCache,
+		BlockProvider
+	}
+
+	/// <summary>Result object describing if/how object was downloaded using the block downloading service.</summary>
+	public interface IResult { }
+
+	/// <summary>Denotes a failure of getting a block.</summary>
+	public interface IFailureResult : IResult { }
+
+	private IFileSystemBlockRepository FileSystemBlockRepository { get; }
+
+	/// <remarks>Implementation must provide caching functionality - i.e. once a block is downloaded, it must be readily available next time it is requested.</remarks>
+	private IBlockProvider BlockProvider { get; }
+	private int MaximumParallelTasks { get; }
+	private SemaphoreSlim SynchronizationRequestsSemaphore { get; } = new(initialCount: 0, maxCount: 1);
+
+	/// <summary>Block hashes that are to be downloaded. Block height represents priority of the priority queue.</summary>
+	/// <remarks>
+	/// Guarded by <see cref="Lock"/>.
+	/// <para>Internal for testing purposes.</para>
+	/// </remarks>
+	internal PriorityQueue<Request, Priority> BlocksToDownload { get; } = new(Priority.Comparer);
+
+	/// <remarks>Guards <see cref="BlocksToDownload"/>.</remarks>
+	private object Lock { get; } = new();
+
+	/// <summary>
+	/// Add a block hash to the queue to be downloaded.
+	/// </summary>
+	public TaskCompletionSource<IResult> Enqueue(uint256 blockHash, Priority priority) =>
```

Comment:
We just need the Task here, don't we?

Reply:
Let me know if you manage to return the `Task` directly without creating warnings for when we don't want to `await` (after `IBlockProvider` returns `null`), I didn't manage to. Also, I had problem with this code because I was breaking the task scheduler because I was starting a thousand of PreProcessing tasks (`MatchAny`) at the same time, which was completely destroying performances of the code.